keps/sig-cli/kubectl-unix-socket.md

```diff
@@ -0,0 +1,177 @@
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Supporting unix socket for `kubectl` is a new feature to provide non-stored credentials for kubeconfig. 
+
+## Motivation
+
+Currently kubectl does not expose a way to send the request over a unix socket. `Curl` does provide a `--unix-socket` option,
+as do many other cli utilities.
+
+Currently several kinds of credentials (passwords, tokens, x509 certificates and etc) are stored in a kubeconfig file and
+persisted on disk. It is not safe and controllable, especially this kubeconfig gets spread to everywhere. Moreover,
+it will become difficult to revoke the access to the apiserver.
+
+Unix sockets are used as a secured proxy into some infrastructure and it would be great to be able to send requests
+from kubectl over this unix socket.
+
+### Goals
+
+* kubectl users can use unix socket to connect to the apiserver:
+    * Add an experimental flag `--experimental-unix-socket` for kubectl, much like `--unix-socket` in `curl`;
+    * Allow specifying unix sockets in kubeconfig file;
+
+### Non-Goals
+
+* Convey credentials through unix sockets;
+* Adopting [SGX] to secure the memory access;
+
+[SGX]: https://www.intel.com/content/www/us/en/architecture-and-technology/software-guard-extensions.html
+
+## Proposal
+
+### User Stories
+
+Issuing commands to any kube api server whereby unix socket is the preferred way to access it.
+
+```bash
+kubectl --experimental-unix-socket ~/.myunixsocket --server mykubecluster.com get pods
+```
+
+### Implementation Details/Notes/Constraints
+
+A `--unix-socket` flag will be added to `kubectl`, much like the --unix-socket in `curl`. The curl docs state: (HTTP)
+Connect through this Unix domain socket, instead of using the network. This is precisely what the
+`kubectl --unix-socket` flag will be used for.
+
+```go
+// staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go
+type ConfigFlags struct {
+    ...
+    
+    // config flags
+    ...
+    Username         *string
+    Password         *string
+    Timeout          *string
+    UnixSocket       *string // +experimental
+    
+    ...
+}
+
+// AddFlags binds client configuration flags to a given flagset
+func (f *ConfigFlags) AddFlags(flags *pflag.FlagSet) {
+    ...
+    if f.UnixSocket != nil {
+        flags.StringVar(f.UnixSocket, "--experimental-unix-socket", *f.UnixSocket, "Path to the unix socket file to use for CLI requests.")
+    }
+}
+```
+
+Also a new field will be appended to `AuthInfo` struct to support specifying unix sockets in a kubeconfig file.
+
+```go
+// staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go
+type AuthInfo struct {
+    ... 
+    // UnixSocket is the path to a unix socket file for connecting. 
+    //+experimental 
+    //+optional
+    UnixSocket string `json:"unix-socket,omitempty"`
```

Comment:
this seems more related to the cluster connection than to the authentication info... would it make sense to use a  Cluster#Server field value of `unix:///path/to/socket`?

Reply:
@liggitt Updated. PTAL.